nototools/emoji_ordering.py

```diff
@@ -0,0 +1,269 @@
+#!/usr/bin/python
+#
+# Copyright 2016 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Represents an ordering of emoji strings, broken into categories and
+subcategories.  Emoji strings are sequences of emoji and related characters
+defined by Unicode, for example flags from pairs of regional indicator symbols,
+varieties of families and couples, emoji with skin tone variants, and
+standalone emoji.  API is provided to return the categories and subcategories,
+the emoji strings in them, and the category, subcategory tuple for an emoji
+string.
+
+Functions are provided to parse two kinds of source data files.
+
+.csv file:
+This has three columns: Category name, emoji-strings separated by space, and
+subcategory name.  The first row, headers, starts with a # and is ignored.
+
+emojiOrdering.txt file:
+This data is a bit raw, so gets validated during processing.
+For an example see the unicodetools svn repo, i.e.
+http://www.unicode.org/utility/trac/browser/trunk/unicodetools/data/emoji/3.0/source/emojiOrdering.txt"""
+
+import argparse
+import codecs
+import collections
+from itertools import chain
+from os import path
+import re
+import sys
+
+def _generate_emoji_to_cat_tuple(category_odict):
```

Comment:
Could.  The line between what's in the class and what's not is a little fuzzy.  The data structure, for example, is built outside the class by the parsing code.  The conversion of the data structure to the inverse map seemed to me more like a utility function that could be used elsewhere (e.g. to verify the original data structure doesn't contain multiple mappings for the same sequence, which I currently don't do) so I left it out of the class.  I'll probably leave it as is.